src/Compression/CompressionCodecDeflateQpl.cpp

```diff
@@ -0,0 +1,412 @@
+            LOG_WARNING(log, "Initialization of hardware-assisted DeflateQpl codec failed, falling back to software DeflateQpl codec. Please check if Intel In-Memory Analytics Accelerator (IAA) is properly set up. QPL Version: {}.",qpl_version);
+            return;
+        }
+        hw_job_ptr_pool[index] = qpl_job_ptr;
+        unLockJob(index);
+    }
+
+    job_pool_ready = true;
+    LOG_DEBUG(log, "Hardware-assisted DeflateQpl codec is ready! QPL Version: {}",qpl_version);
+}
+
+DeflateQplJobHWPool::~DeflateQplJobHWPool()
+{
+    for (UInt32 i = 0; i < MAX_HW_JOB_NUMBER; ++i)
+    {
+        if (hw_job_ptr_pool[i])
+        {
+            while (!tryLockJob(i));
+            qpl_fini_job(hw_job_ptr_pool[i]);
+            unLockJob(i);
+            hw_job_ptr_pool[i] = nullptr;
+        }
+    }
+    job_pool_ready = false;
+}
+
+qpl_job * DeflateQplJobHWPool::acquireJob(UInt32 &job_id)
+{
+    if (isJobPoolReady())
+    {
+        UInt32 retry = 0;
+        auto index = distribution(random_engine);
+        while (!tryLockJob(index))
+        {
+            index = distribution(random_engine);
+            retry++;
+            if (retry > MAX_HW_JOB_NUMBER)
+            {
+                return nullptr;
+            }
+        }
+        job_id = MAX_HW_JOB_NUMBER - index;
+        assert(index < MAX_HW_JOB_NUMBER);
+        return hw_job_ptr_pool[index];
+    }
+    else
+        return nullptr;
+}
+
+void DeflateQplJobHWPool::releaseJob(UInt32 job_id)
+{
+    if (isJobPoolReady())
+        unLockJob(MAX_HW_JOB_NUMBER - job_id);
+}
+
+bool DeflateQplJobHWPool::tryLockJob(UInt32 index)
+{
+    bool expected = false;
+    assert(index < MAX_HW_JOB_NUMBER);
+    return hw_job_ptr_locks[index].compare_exchange_strong(expected, true);
+}
+
+ void DeflateQplJobHWPool::unLockJob(UInt32 index)
+ {
+    assert(index < MAX_HW_JOB_NUMBER);
+    hw_job_ptr_locks[index].store(false);
+ }
+
+//HardwareCodecDeflateQpl
+HardwareCodecDeflateQpl::HardwareCodecDeflateQpl()
+    :log(&Poco::Logger::get("HardwareCodecDeflateQpl"))
+{
+}
+
+HardwareCodecDeflateQpl::~HardwareCodecDeflateQpl()
+{
+#ifndef NDEBUG
+    assert(decomp_async_job_map.empty());
+#else
+    if (!decomp_async_job_map.empty())
+    {
+        LOG_WARNING(log, "Find un-released job when HardwareCodecDeflateQpl destroy");
+        for (auto it : decomp_async_job_map)
+        {
+            DeflateQplJobHWPool::instance().releaseJob(it.first);
+        }
+        decomp_async_job_map.clear();
+    }
+#endif
+}
+
+Int32 HardwareCodecDeflateQpl::doCompressData(const char * source, UInt32 source_size, char * dest, UInt32 dest_size) const
+{
+    UInt32 job_id = 0;
+    qpl_job* job_ptr = nullptr;
+    UInt32 compressed_size = 0;
+    if (!(job_ptr = DeflateQplJobHWPool::instance().acquireJob(job_id)))
+    {
+        LOG_WARNING(log, "DeflateQpl HW codec failed, falling back to SW codec.(Details: doCompressData->acquireJob fail, probably job pool exhausted)");
```

Comment:
While looking at the code for a final check, I noticed that we'll log a warning (LOG_WARNING) when the job pool is exhausted. This situation may occur quite often (depending on the workload) and it's internally handled ... should the log level be demoted to INFO to not spam the user? --> See l. 148, 181, 216. 

(on the other hand, the log messages in l. 168, 196, 236 use level WARNING for good reason)

Reply:
Done